### PR TITLE
[Merged by Bors] - feat: port Data.PNat.Xgcd

### PR DESCRIPTION
**Pull request by qawbecrdtey**
Initial files generated by `start_port.sh` has been moved from `Data/Pnat` to `Data/PNat`.

### CODE REVIEW

File: Mathlib/Data/PNat/Xgcd.lean [hunk 1]
```diff
@@ -62,12 +62,9 @@ instance : SizeOf XgcdType :=
...
   ⟨fun u =>
-    "[[[" ++ repr (u.wp + 1) ++ ", " ++ repr u.x ++ "], [" ++ repr u.y ++ ", " ++ repr (u.zp + 1) ++
-              "]], [" ++
-            repr (u.ap + 1) ++
-          ", " ++
-        repr (u.bp + 1) ++
-      "]]"⟩
+    "[[[" ++ repr (u.wp + 1) ++ ", " ++ repr u.x ++ "], [" ++
+      repr u.y ++ ", " ++ repr (u.zp + 1) ++ "]], [" ++
+      repr (u.ap + 1) ++ ", " ++ repr (u.bp + 1) ++ "]]"⟩
```
Comment (qawbecrdtey) [inline Mathlib/Data/PNat/Xgcd.lean:67]: Formatting required

File: Mathlib/Data/PNat/Xgcd.lean [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+#align pnat.xgcd_type.is_special' PNat.XgcdType.is_special'
+
+theorem is_special_iff : u.is_special ↔ u.is_special' := by
+  dsimp [is_special, is_special']
+  let ⟨wp, x, y, zp, ap, bp⟩ := u
+  constructor <;> intro h <;> simp [w, z, succPNat] at * <;>
+    simp only [← coe_inj, mul_coe, mk_coe] at *
+  . simp_all [← h, Nat.mul, Nat.succ_eq_add_one]; ring
+  . simp [Nat.succ_eq_add_one, Nat.mul_add, Nat.add_mul, ← Nat.add_assoc] at h; rw [← h]; ring
+  -- Porting note: Below doesn't work, so I fixed it as above.
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:160]: I think you can delete the entire commented code below.
Your new proof is so much shorter that I don't think anybody is interested in reviving the old proof.

File: Mathlib/Data/PNat/Xgcd.lean [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+  --   rw [Nat.succ_add u.wp, Nat.succ_add] at h
+  --   exact h
+
+#align pnat.xgcd_type.is_special_iff PNat.XgcdType.is_special_iff
+
+/-- `is_reduced` holds if the two entries in the vector are the
+ same.  The reduction algorithm will produce a system with this
+ property, whose product vector is the same as for the original
+ system. -/
+def is_reduced : Prop :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:204]: Should this be
```suggestion
def isReduced : Prop :=
```

File: Mathlib/Data/PNat/Xgcd.lean [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+  u.ap = u.bp
+#align pnat.xgcd_type.is_reduced PNat.XgcdType.is_reduced
+
+/-- `is_reduced'` is an alternative of `is_reduced`. -/
+def is_reduced' : Prop :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:209]:
```suggestion
def isReduced' : Prop :=
```

File: Mathlib/Data/PNat/Xgcd.lean [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+def succ₂ (t : ℕ × ℕ) : ℕ × ℕ :=
+  ⟨t.1.succ, t.2.succ⟩
+#align pnat.xgcd_type.succ₂ PNat.XgcdType.succ₂
+
+theorem v_eq_succ_vp : u.v = succ₂ u.vp := by
+  ext <;> dsimp [v, vp, w, z, a, b, succ₂] <;> (repeat' rw [Nat.succ_eq_add_one]; ring_nf)
+#align pnat.xgcd_type.v_eq_succ_vp PNat.XgcdType.v_eq_succ_vp
+
+/-- `is_special` holds if the matrix has determinant one. -/
+def is_special : Prop :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:144]:
```suggestion
def isSpecial : Prop :=
```

File: Mathlib/Data/PNat/Xgcd.lean [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+  u.wp + u.zp + u.wp * u.zp = u.x * u.y
+#align pnat.xgcd_type.is_special PNat.XgcdType.is_special
+
+/-- `is_special'` is an alternative of `is_special`. -/
+def is_special' : Prop :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:149]:
```suggestion
def isSpecial' : Prop :=
```

**Review by jcommelin — COMMENTED** [review]
If I'm not mistaken, when you define a `Prop` the name should also be camelCase. Please fix.

File: Mathlib/Data/PNat/Xgcd.lean (hunk 3 above)
Comment (qawbecrdtey) [reply Mathlib/Data/PNat/Xgcd.lean:204]: Should it be? According to [naming convention](https://github.com/leanprover-community/mathlib4/wiki/Porting-wiki#naming-convention), This is a term of `Prop`, which means this should be written as `is_reduced`, not `isReduced`.

File: Mathlib/Data/PNat/Xgcd.lean (hunk 4 above)
Comment (qawbecrdtey) [reply Mathlib/Data/PNat/Xgcd.lean:209]: Same as above

File: Mathlib/Data/PNat/Xgcd.lean (hunk 5 above)
Comment (qawbecrdtey) [reply Mathlib/Data/PNat/Xgcd.lean:144]: Same as above

File: Mathlib/Data/PNat/Xgcd.lean (hunk 3 above)
Comment (qawbecrdtey) [reply Mathlib/Data/PNat/Xgcd.lean:204]: Thinking about it again, it seems that I'm the one who's mistaken. I'll fix it.

File: Mathlib/Data/PNat/Xgcd.lean [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+def isSpecial : Prop :=
+  u.wp + u.zp + u.wp * u.zp = u.x * u.y
+#align pnat.xgcd_type.is_special PNat.XgcdType.isSpecial
+
+/-- `isSpecial'` is an alternative of `isSpecial`. -/
+def isSpecial' : Prop :=
+  u.w * u.z = succPNat (u.x * u.y)
+#align pnat.xgcd_type.is_special' PNat.XgcdType.isSpecial'
+
+theorem is_special_iff : u.isSpecial ↔ u.isSpecial' := by
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:153]: Please fix this name. If you use VSCode, do the following:

1. select `is_special`
2. run the following loop:
  - hit `Ctrl-D` to select the next occurence
  - if the occurence should be skipped, hit `Ctrl-K` to "kill" that selection
  - (if necessary, hit `Ctrl-U` to backtrack)
3. after you have selected all occurences of `is_special` that need to be changed, make the edit to `isSpecial`.

File: Mathlib/Data/PNat/Xgcd.lean [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+def isReduced : Prop :=
+  u.ap = u.bp
+#align pnat.xgcd_type.is_reduced PNat.XgcdType.isReduced
+
+/-- `isReduced'` is an alternative of `isReduced`. -/
+def isReduced' : Prop :=
+  u.a = u.b
+#align pnat.xgcd_type.is_reduced' PNat.XgcdType.isReduced'
+
+theorem is_reduced_iff : u.isReduced ↔ u.isReduced' :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:177]: Same for `is_reduced`.

File: Mathlib/Data/PNat/Xgcd.lean [hunk 9]
```diff
@@ -0,0 +1,556 @@
...
+  rw [reduce]
+  exact if_pos h
+#align pnat.xgcd_type.reduce_a PNat.XgcdType.reduce_a
+
+theorem reduce_b {u : XgcdType} (h : u.r ≠ 0) : u.reduce = u.step.reduce.flip := by
+  rw [reduce]
+  exact if_neg h
+#align pnat.xgcd_type.reduce_b PNat.XgcdType.reduce_b
+
+theorem reduce_reduced : ∀ u : XgcdType, u.reduce.isReduced
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:362]:
```suggestion
theorem reduce_isReduced : ∀ u : XgcdType, u.reduce.isReduced
```

File: Mathlib/Data/PNat/Xgcd.lean [hunk 10]
```diff
@@ -0,0 +1,556 @@
...
+      (fun h => by
+        rw [reduce_a h]
+        exact u.finish_isReduced)
+      fun h => by
+      have : SizeOf.sizeOf u.step < SizeOf.sizeOf u := u.step_wf h
+      rw [reduce_b h, flip_isReduced]
+      apply reduce_reduced
+#align pnat.xgcd_type.reduce_reduced PNat.XgcdType.reduce_reduced
+
+theorem reduce_reduced' (u : XgcdType) : u.reduce.isReduced' :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:374]:
```suggestion
theorem reduce_isReduced' (u : XgcdType) : u.reduce.isReduced' :=
```

File: Mathlib/Data/PNat/Xgcd.lean [hunk 11]
```diff
@@ -0,0 +1,556 @@
...
+  (isReduced_iff _).mp u.reduce_reduced
+#align pnat.xgcd_type.reduce_reduced' PNat.XgcdType.reduce_reduced'
+
+theorem reduce_special : ∀ u : XgcdType, u.isSpecial → u.reduce.isSpecial
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:378]:
```suggestion
theorem reduce_isSpecial : ∀ u : XgcdType, u.isSpecial → u.reduce.isSpecial
```

File: Mathlib/Data/PNat/Xgcd.lean [hunk 12]
```diff
@@ -0,0 +1,556 @@
...
+      (fun h hs => by
+        rw [reduce_a h]
+        exact u.finish_isSpecial hs)
+      fun h hs => by
+      have : SizeOf.sizeOf u.step < SizeOf.sizeOf u := u.step_wf h
+      rw [reduce_b h]
+      exact (flip_isSpecial _).mpr (reduce_special _ (u.step_isSpecial hs))
+#align pnat.xgcd_type.reduce_special PNat.XgcdType.reduce_special
+
+theorem reduce_special' (u : XgcdType) (hs : u.isSpecial) : u.reduce.isSpecial' :=
```
Comment (jcommelin) [inline Mathlib/Data/PNat/Xgcd.lean:390]:
```suggestion
theorem reduce_isSpecial' (u : XgcdType) (hs : u.isSpecial) : u.reduce.isSpecial' :=
```

**Comment by jcommelin** [comment]
Four more naming issues. After those are fixed, LGTM.

bors d+

**Pull request closed without merging**